haplotype_tracking.md

```diff
@@ -0,0 +1,442 @@
+        pa.add_segment(a, b)
+
+for j in ancestry:
+    print(j, ancestry[j].segments)
+```
+
+Here we see that the ancestral segments above nodes 2, 9 and 10
+span the entire genome, while node 6 only spans part of the genome.
+Since this tree sequence was simulated by msprime using
+`record_full_arg=True`, interpretation is straightforward:
+at 300 generations ago, there were three extant genomes
+from which the samples inherited, and the inherited segments are
+as listed here.
+Note that this does not mean that "node 2 was laive 300 generations ago"
+(clearly, as node 2 represetns an extant, sampled genome),
+but rather that there are no other ancestral genomes recorded explicitly
+in the tree sequence that lie on the path along
+which node 2 has inherited it's genome.
+
+We can use this to plot the state of the process as we move back through time.
+Here are line segments depicting which bits of the genome are inherited
+from which others at five times in the past.
+
+```{code-cell} ipython3
+import matplotlib.pyplot as plt
+from matplotlib import collections as mc
+
+times = [0, 50, 100, 200, 1000]
+fig, axes = plt.subplots(len(times), figsize=(6, 8))
+
+ancestry = {n.id : SegmentList() for n in ts.nodes()}
+for j in ts.samples():
+    ancestry[j].add_segment(0, ts.sequence_length)
+
+k = 0
+for e in ts.edges():
+    t = ts.node(e.parent).time
+    if t > times[k]:
+        ax = axes[k]
+        ax.set_title(f"t = {times[k]}")
+        ax.set_xlim(0, ts.sequence_length)
+        ax.set_ylim(0, ts.num_nodes)
+        lc = mc.LineCollection(
+            [(a, j), (b, j)] for j in ancestry
+            for a, b in ancestry[j].segments
+        )
+        ax.add_collection(lc)
+        k += 1
+        if k == len(times):
+            break
+    ca = ancestry[e.child]
+    pa = ancestry[e.parent]
+    seg = ca.remove_segment(e.left, e.right)
+    for a, b in seg.segments:
+        pa.add_segment(a, b)
+
+plt.tight_layout()
+```
+
+## Tracking ancestral segments
+
+The code above gives an example of the basic algorithm that updates inherited segments.
+However, for most purposes we want some additional structure.
+For instance, let's say we'd like to know how many samples inherit from each segment.
+To do this, we use a structure similar to the above,
+but each segment now has a *label*, that is just the number of samples inheriting from it.
+We'll demonstrate with a slightly more complex example tree sequence:
+
+```{code-cell} ipython3
+:tags: [remove-cell]
+def sim_example2():
+    ts = msprime.sim_ancestry(6, sequence_length=7e5,
+            recombination_rate=1e-8, population_size=1000,
+            random_seed=21)
+    ts.dump("data/haplotypes2.trees")
+
+sim_example2()  # uncomment to recreate the tree seqs used in this notebook
+```
+
+```{code-cell} ipython3
+ts = tskit.load("data/haplotypes2.trees")
+ts
+```
+
+Here is a data structure for a list of segments with labels:
+
+```{code-cell} ipython3
+class LabelSegmentList:
```

Comment:
It might be worth considering inheriting from ``collections.abc.MutableSequence`` here, as this would give you all the dunder methods that you're implementing. I think these might be a bit scary to non-python people, and are a bit of a distraction from the main point.

Reply:
I guess you'd need to inherit from ``list`` then to get the actual storage. Maybe that's OK?